<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Crispulator Internals · Crispulator.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script><script src="https://analytics.tamasnagy.com/js/script.js" data-domain="tamasnagy.com" defer></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="Crispulator.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">Crispulator.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../tutorial/">Tutorial</a></li><li><a class="tocitem" href="../custom/">Custom Simulations</a></li><li><a class="tocitem" href="../multiprocessing/">Multiprocessing</a></li><li class="is-active"><a class="tocitem" href>Crispulator Internals</a><ul class="internal"><li><a class="tocitem" href="#Contents"><span>Contents</span></a></li><li><a class="tocitem" href="#Index"><span>Index</span></a></li><li><a class="tocitem" href="#Simulation-Types"><span>Simulation Types</span></a></li><li><a class="tocitem" href="#Key-functions"><span>Key functions</span></a></li><li><a class="tocitem" href="#Miscellaneous-Types"><span>Miscellaneous Types</span></a></li><li><a class="tocitem" href="#Miscellaneous-functions"><span>Miscellaneous functions</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Crispulator Internals</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Crispulator Internals</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/tlnagy/Crispulator.jl/blob/master/docs/src/internals.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Crispulator-Internals"><a class="docs-heading-anchor" href="#Crispulator-Internals">Crispulator Internals</a><a id="Crispulator-Internals-1"></a><a class="docs-heading-anchor-permalink" href="#Crispulator-Internals" title="Permalink"></a></h1><p>More in-depth documentation on specific types and functions.</p><h2 id="Contents"><a class="docs-heading-anchor" href="#Contents">Contents</a><a id="Contents-1"></a><a class="docs-heading-anchor-permalink" href="#Contents" title="Permalink"></a></h2><ul><li><a href="#Crispulator-Internals">Crispulator Internals</a></li><li class="no-marker"><ul><li><a href="#Contents">Contents</a></li><li><a href="#Index">Index</a></li><li><a href="#Simulation-Types">Simulation Types</a></li><li><a href="#Key-functions">Key functions</a></li><li><a href="#Miscellaneous-Types">Miscellaneous Types</a></li><li><a href="#Miscellaneous-functions">Miscellaneous functions</a></li></ul></li></ul><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><ul><li><a href="#Crispulator.Barcode"><code>Crispulator.Barcode</code></a></li><li><a href="#Crispulator.CRISPRi"><code>Crispulator.CRISPRi</code></a></li><li><a href="#Crispulator.CRISPRn"><code>Crispulator.CRISPRn</code></a></li><li><a href="#Crispulator.Cas9Behavior"><code>Crispulator.Cas9Behavior</code></a></li><li><a href="#Crispulator.Delta"><code>Crispulator.Delta</code></a></li><li><a href="#Crispulator.FacsScreen"><code>Crispulator.FacsScreen</code></a></li><li><a href="#Crispulator.GrowthScreen"><code>Crispulator.GrowthScreen</code></a></li><li><a href="#Crispulator.KDPhenotypeRelationship"><code>Crispulator.KDPhenotypeRelationship</code></a></li><li><a href="#Crispulator.Library"><code>Crispulator.Library</code></a></li><li><a href="#Crispulator.auprc"><code>Crispulator.auprc</code></a></li><li><a href="#Crispulator.auroc"><code>Crispulator.auroc</code></a></li><li><a href="#Crispulator.construct_library"><code>Crispulator.construct_library</code></a></li><li><a href="#Crispulator.counts_to_freqs"><code>Crispulator.counts_to_freqs</code></a></li><li><a href="#Crispulator.differences_between_bins"><code>Crispulator.differences_between_bins</code></a></li><li><a href="#Crispulator.linear"><code>Crispulator.linear</code></a></li><li><a href="#Crispulator.noise"><code>Crispulator.noise</code></a></li><li><a href="#Crispulator.select"><code>Crispulator.select</code></a></li><li><a href="#Crispulator.sequencing"><code>Crispulator.sequencing</code></a></li><li><a href="#Crispulator.sigmoid"><code>Crispulator.sigmoid</code></a></li><li><a href="#Crispulator.signal"><code>Crispulator.signal</code></a></li><li><a href="#Crispulator.transfect"><code>Crispulator.transfect</code></a></li><li><a href="#Crispulator.venn"><code>Crispulator.venn</code></a></li></ul><h2 id="Simulation-Types"><a class="docs-heading-anchor" href="#Simulation-Types">Simulation Types</a><a id="Simulation-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Simulation-Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Crispulator.FacsScreen" href="#Crispulator.FacsScreen"><code>Crispulator.FacsScreen</code></a> — <span class="docstring-category">Type</span></header><section><div><p>A type representing the parameters used in a typical FACS screen.</p><ul><li><p><code>num_genes</code></p><p>Number of genes targeted by the screen</p></li><li><p><code>coverage</code></p><p>Number of guides per gene</p></li><li><p><code>representation</code></p><p>Number of cells with each guide. <code>representation=10</code> means that there are 10 times as many cells as guides during transfection. The actual number of cells per guide post-transfection will be less depending on the MOI</p></li><li><p><code>moi</code></p><p>The multiplicity of infection, <span>$\lambda$</span>, of the screen. We model this as a Poisson process during transfection (see <a href="#Crispulator.transfect"><code>Crispulator.transfect</code></a>).</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>We <strong>do not</strong> model multiple infections. We assume that the MOI is properly selected and less than half the cells are transfected by any virus, i.e. <span>$\lambda \lt 0.5$</span> and then select only the cells that have a single transfection occurrence:</p><p class="math-container">\[P(x = 1; Poisson(λ))\]</p><p>For <span>$\lambda = 0.25$</span> this works out to being <span>$\approx 19.5\%$</span> of the number of cells (<code>num_genes</code> * <code>coverage</code> * <code>representation</code>).</p></div></div></li></ul><ul><li><p><code>σ</code></p><p>The standard deviation expected for cells during FACS sorting. This should be set according to the biological variance experimentally observed, e.g. in the fluorescence intensity of isogenic cells</p></li><li><p><code>bin_info</code></p><p>Range of guide phenotypes to collect in each bin</p><h1>Example</h1><p>In the following example</p><pre><code class="language-julia hljs">p = 0.05
bin_info = Dict(:bin1 =&gt; (0.0, p), :bin2 =&gt; (1.0-p, 1.0))</code></pre><p>The 5th percentile of cells sorted according to their phenotype (fluorescence, size, etc) will be compared to the 95th percentile.</p></li></ul><ul><li><p><code>bottleneck_representation</code></p><p>Number of cells sorted expressed as an integer multiple of the number of guides</p></li><li><p><code>seq_depth</code></p><p>Sequencing depth as a integer multiple of the number of guides, i.e. <code>seq_depth=10</code> is equivalent to <code>10 * num_genes * coverage</code> reads.</p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/common.jl#L51-L55">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.GrowthScreen" href="#Crispulator.GrowthScreen"><code>Crispulator.GrowthScreen</code></a> — <span class="docstring-category">Type</span></header><section><div><p>A type representing the parameters used in a typical growth-based screen.</p><ul><li><p><code>num_genes</code></p><p>Number of genes targeted by the screen</p></li><li><p><code>coverage</code></p><p>Number of guides per gene</p></li><li><p><code>representation</code></p><p>Number of cells with each guide. <code>representation=10</code> means that there are 10 times as many cells as guides during transfection. The actual number of cells per guide post-transfection will be less depending on the MOI</p></li><li><p><code>moi</code></p><p>The multiplicity of infection, <span>$\lambda$</span>, of the screen. We model this as a Poisson process during transfection (see <a href="#Crispulator.transfect"><code>Crispulator.transfect</code></a>).</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>We <strong>do not</strong> model multiple infections. We assume that the MOI is properly selected and less than half the cells are transfected by any virus, i.e. <span>$\lambda \lt 0.5$</span> and then select only the cells that have a single transfection occurrence:</p><p class="math-container">\[P(x = 1; Poisson(λ))\]</p><p>For <span>$\lambda = 0.25$</span> this works out to being <span>$\approx 19.5\%$</span> of the number of cells (<code>num_genes</code> * <code>coverage</code> * <code>representation</code>).</p></div></div></li></ul><ul><li><p><code>seq_depth</code></p><p>Sequencing depth as a integer multiple of the number of guides, i.e. <code>seq_depth=10</code> is equivalent to <code>10 * num_genes * coverage</code> reads.</p></li></ul><ul><li><p><code>bottleneck_representation</code></p><p>For growth screens, how much of a bottleneck is applied. This is minimum number of cells that is kept when passaging the pool of cells. This is expressed as an integer multiple of the number of guides.</p></li><li><p><code>num_bottlenecks</code></p><p>For growth screens, how many bottlenecks are applied. This is the integer number of passages during the growth screen.</p></li><li><p><code>noise</code></p><p>Before each passage, the theoretical phenotype of each cell is convolved with a normal noise distribution with a standard deviation, σ, dictated by this parameter. This should be set based on an expectation of noisiness of the subsampling</p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/common.jl#L121-L125">source</a></section></article><h2 id="Key-functions"><a class="docs-heading-anchor" href="#Key-functions">Key functions</a><a id="Key-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Key-functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Crispulator.construct_library" href="#Crispulator.construct_library"><code>Crispulator.construct_library</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">construct_library(setup, lib)
</code></pre><p>Constructs the guide library for <code>N</code> genes with <code>coverage</code> number of guides per gene. Returns a tuple of guides and their relative frequencies (assigned randomly).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/library.jl#L242">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.transfect" href="#Crispulator.transfect"><code>Crispulator.transfect</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">transfect(setup, lib, guides, guide_freqs_dist)
</code></pre><p>Simulates the transfection the of the library of guides (<code>guides</code>) according to the parameters in <a href="#Crispulator.Library"><code>Crispulator.Library</code></a> <code>lib</code> and the frequencies represented by the Categorical distribution <code>guide_freqs_dist</code>.</p><p>This function creates a population of cells post-transfection where each cell has single guide RNA present according to the frequencies of the guides in the present in the library. The cells are then grown up so that there are <code>setup.bottleneck_representation * num_guides</code> of them. For FACS screens we assume a simple linear expansion of the cells, while for Growth screens the cells grow according as a function of their phenotype and degree of knockdown.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/transfection.jl#L30">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.select" href="#Crispulator.select"><code>Crispulator.select</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">select(setup, cells, cell_phenotypes, guides; debug)
</code></pre><p>Given <code>cells</code>, a vector of integers, and <code>guides</code>, a vector of barcodes, performs simulated FACS sorting of the cells into <code>bins</code> with the given cutoffs. <code>σ</code> refers to the spread of the phenotype during the FACS screen.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/selection.jl#L2">source</a></section><section><div><pre><code class="language-julia hljs">select(setup, initial_cells, initial_cell_phenotypes, guides)
</code></pre><p>Given a population <code>initial_cells</code> with <code>initial_cell_phenotypes</code>, this function models the selection step of a <a href="@ref">GrowthScreen</a>. Cells are allowed to double according to their growth phenotype and then the population is subsampled to simulate passaging to maintain the <code>bottleneck_representation</code> as defined in <code>setup</code>. The number of doublings + subsamplings are controlled by the <code>num_bottlenecks</code> parameter in <code>setup</code>.</p><p>So for a screen where there are three bottlenecks and the <code>length(initial_cells) == bottleneck_representation</code> then we would expect the following behavior:</p><pre><code class="nohighlight hljs">           Cell population</code></pre><p>┌────────────────────────────────────────┐ 2 │           .-           .r|          ..|│   │         r<em>`|         _-&#39; |        .r/ |│   │       .&quot;   |       .</em>    |      .r&#39;   |│   │    .-&quot;      |    ,&quot;<code>|    ,/&#39;     |│   │  _-</code>        | .-/        |  .<em>`       |│ 1 │_</em>           Lr&#39;          |r/          |│   └────────────────────────────────────────┘    0                                      3                  Bottleneck #</p><p>If the coverage of the library by <code>initial_cells</code> is less than the <code>bottleneck_representation</code> than the population will be allowed to double until passaging is needed or the <code>num_bottlenecks</code> is hit, which ever comes first.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/selection.jl#L55">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.sequencing" href="#Crispulator.sequencing"><code>Crispulator.sequencing</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">sequencing(depths, guides, samples)
</code></pre><p>Simulates next-gen sequencing by generating a Categorical distribution based on frequencies of each guide in each bin and then randomly samples from this distributions up to the depth provided in <code>depths</code> for each bin. Returns a dict of DataFrames with the bin names as the keys. This object can then be passed through <a href="#Crispulator.counts_to_freqs"><code>Crispulator.counts_to_freqs</code></a> followed by <a href="#Crispulator.differences_between_bins"><code>Crispulator.differences_between_bins</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/sequencing.jl#L16">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.counts_to_freqs" href="#Crispulator.counts_to_freqs"><code>Crispulator.counts_to_freqs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">counts_to_freqs(bin_cells, guide_count)
</code></pre><p>Converts raw counts to frequencies by dividing by the total number of reads for each sample.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/sequencing.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.differences_between_bins" href="#Crispulator.differences_between_bins"><code>Crispulator.differences_between_bins</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">differences_between_bins(raw_data)
</code></pre><p>Given the raw data from <a href="#Crispulator.sequencing"><code>Crispulator.sequencing</code></a> returns two DataFrames</p><ol><li><p><code>guide_data</code>: This DataFrame contains the per-guide level data including the  log2 fold change in the normalized frequencies of each guide between each  pairwise combination of bins. Thus, if there are <code>n</code> bins, then it computes  the log2 fold changes for the <span>$\frac{n!}{2(n-2)!}$</span> combinations</p></li><li><p><code>gene_data</code>: This DataFrame contains the same information but grouped by  gene. The log2 fold change data from the first DataFrame is used to calculate  the average log2 fold change per gene and a pvalue computed using a  <a href="https://en.wikipedia.org/wiki/Mann-Whitney_U_test">Mann-Whitney U-test</a> as  measure of how consistently shifted the guides are of this gene versus the  population of negative control guides. (see below for more info)</p></li></ol><p>A typical 2 bin <code>guide_data</code> DataFrame contains the following columns:</p><table><tr><th style="text-align: left">Column Name</th><th style="text-align: left">Meaning</th></tr><tr><td style="text-align: left"><code>gene</code></td><td style="text-align: left">the gene ID of that this guide targets</td></tr><tr><td style="text-align: left"><code>knockdown</code></td><td style="text-align: left">activity of the guide on 0 to 1 scale, where 1 is complete knockout</td></tr><tr><td style="text-align: left"><code>barcodeid</code></td><td style="text-align: left">the ID of this specific guide</td></tr><tr><td style="text-align: left"><code>theo_phenotype</code></td><td style="text-align: left">expected phenotype of this guide, generally a -1 to 1 scale</td></tr><tr><td style="text-align: left"><code>behavior</code></td><td style="text-align: left">whether the target gene displays a linear or sigmoidal response to incomplete knockdown (see <a href="#Crispulator.Library"><code>Crispulator.Library</code></a> for more details)</td></tr><tr><td style="text-align: left"><code>class</code></td><td style="text-align: left">which phenotype distribution the target gene was drawn from (see <a href="#Crispulator.Library"><code>Crispulator.Library</code></a> for more details). Serves as the &quot;ground truth&quot; label against which screen performance is evaluated, e.g. with <a href="#Crispulator.auprc"><code>Crispulator.auprc</code></a></td></tr><tr><td style="text-align: left"><code>initial_freq</code></td><td style="text-align: left">frequency of guide post-transfection (see <a href="#Crispulator.transfect"><code>Crispulator.transfect</code></a>)</td></tr><tr><td style="text-align: left"><code>counts_bin1</code></td><td style="text-align: left">the raw number of reads for each guide in the first bin</td></tr><tr><td style="text-align: left"><code>freqs_bin1</code></td><td style="text-align: left">the number of reads for each guide divided by the total number of reads in this bin</td></tr><tr><td style="text-align: left"><code>rel_freqs_bin1</code></td><td style="text-align: left">the frequency of each guide divided by the median frequency of negative control guides</td></tr><tr><td style="text-align: left"><code>counts_bin2</code></td><td style="text-align: left">the raw number of reads for each guide in the second bin</td></tr><tr><td style="text-align: left"><code>freqs_bin2</code></td><td style="text-align: left">the number of reads for each guide divided by the total number of reads in this bin</td></tr><tr><td style="text-align: left"><code>rel_freqs_bin2</code></td><td style="text-align: left">the frequency of each guide divided by the median frequency of negative control guides for this bin</td></tr><tr><td style="text-align: left"><code>log2fc_bin2_div_bin1</code></td><td style="text-align: left">the log2 fold change in relative guide frequencies between <code>bin2</code> and <code>bin1</code>, equivalent to <code>log2(rel_freqs_bin2/rel_freqs_bin1)</code></td></tr></table><p>A typical <code>gene_data</code> DataFrame contains the following data:</p><table><tr><th style="text-align: left">Column Name</th><th style="text-align: left">Meaning</th></tr><tr><td style="text-align: left"><code>gene</code></td><td style="text-align: left">this gene&#39;s ID</td></tr><tr><td style="text-align: left"><code>class</code></td><td style="text-align: left">see above</td></tr><tr><td style="text-align: left"><code>behavior</code></td><td style="text-align: left">see above</td></tr><tr><td style="text-align: left"><code>mean_bin2_div_bin1</code></td><td style="text-align: left">the mean log 2 fold change in relative frequencies between from <code>bin1</code> to <code>bin2</code> for all the guides targeting this gene. Calculated as <span>$\frac{1}{k}\sum_k \text{log2fc_bin2_div_bin1}_k$</span> for the <span>$k$</span> guides targeting each gene</td></tr><tr><td style="text-align: left"><code>pvalue_bin2_div_bin1</code></td><td style="text-align: left">the -log10 pvalue of the log2 fold changes of all guides targeting this gene as computed by the non-parametric Mann-Whitney U-test. A measure of the consistency of the log 2 fold changes<sup class="footnote-reference"><a id="citeref-1" href="#footnote-1">[1]</a></sup></td></tr><tr><td style="text-align: left"><code>absmean_bin2_div_bin1</code></td><td style="text-align: left">absolute value of <code>mean_bin2_div_bin1</code> per-gene</td></tr><tr><td style="text-align: left"><code>pvalmeanprod_bin2_div_bin1</code></td><td style="text-align: left"><code>mean_bin2_div_bin1</code> multiplied with the <code>pvalue_bin2_div_bin1</code> per-gene</td></tr></table><p><strong>Further reading</strong></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/processing.jl#L1">source</a></section></article><h2 id="Miscellaneous-Types"><a class="docs-heading-anchor" href="#Miscellaneous-Types">Miscellaneous Types</a><a id="Miscellaneous-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Miscellaneous-Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Crispulator.Library" href="#Crispulator.Library"><code>Crispulator.Library</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Wrapper containing all library construction parameters</p><ul><li><p><code>knockdown_dist</code></p><p>Distribution of guide knockdown efficiencies</p></li><li><p><code>knockdown_probs</code></p></li><li><p><code>max_phenotype_dists</code></p><p>Maximum phenotype categories mapped to their probability of being selected and the distribution to draw from if they are selected.</p><h2>Example</h2><p>The basic layout is a <code>Dict</code> mapping a class name to a tuple of the probability of selecting this class and then the <a href="https://juliastats.github.io/Distributions.jl/latest/types.html#Sampleable-1"><code>Distributions.Sampleable</code></a> from which to draw a random phenotype from this class. The probabilities across all the classes should add up to 1.</p><pre><code class="language-julia hljs">max_phenotype_dists = Dict{Symbol, Tuple{Float64, Sampleable}}(
    :inactive =&gt; (0.60, Delta(0.0)),
    :negcontrol =&gt; (0.1, Delta(0.0)),
    :increasing =&gt; (0.3, truncated(Normal(0.1, 0.1), 0.025, 1)),
);
Library(max_phenotype_dists, CRISPRi());</code></pre><p>For example, here we are making three different classes of &quot;genes&quot;: the first group are :inactive, i.e. they have no phenotype, so we&#39;ll set their phenotypes to 0.0 using a <a href="#Crispulator.Delta"><code>Crispulator.Delta</code></a>. We&#39;ll also make them 60% of all the genes. The second group are the negative controls :negcontrol (the only required group) which make up 10% of the population of genes and also have no effect. The final group is :increasing which makes up 30% of all genes and which are represented by a Normal(μ=0.1, σ=0.1) distribution clamped between 0.025 and 1.</p></li></ul><ul><li><p><code>phenotype_probs</code></p></li><li><p><code>kd_phenotype_relationships</code></p><p>Knockdown-phenotype relationships mapped to their probability of being selected and their respective <a href="#Crispulator.KDPhenotypeRelationship"><code>Crispulator.KDPhenotypeRelationship</code></a></p></li></ul><ul><li><p><code>relationship_probs</code></p></li><li><p><code>cas9_behavior</code></p><p>Whether this library is <a href="#Crispulator.CRISPRi"><code>Crispulator.CRISPRi</code></a> or <a href="#Crispulator.CRISPRn"><code>Crispulator.CRISPRn</code></a></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/library.jl#L87-L91">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.Barcode" href="#Crispulator.Barcode"><code>Crispulator.Barcode</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Any entity that is tracked through the pooled experiment. For CRISPR screens, this is equivalent to the sgRNA. This object stores properties relating to the performance of this entity in the screen.</p><ul><li><p><code>gene</code></p><p>The target gene id</p></li><li><p><code>knockdown</code></p><p>The knockdown efficiency</p></li><li><p><code>theo_phenotype</code></p><p>The theoretical phenotype exerted</p></li><li><p><code>obs_phenotype</code></p><p>The observed phenotype (only relevant for FACS screens)</p></li><li><p><code>behavior</code></p><p>The gene behavior, either sigmoidal or linear</p></li><li><p><code>class</code></p><p>The gene activity class, either inactive, up or down</p></li><li><p><code>initial_freq</code></p><p>Initial frequency after transfection</p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/common.jl#L2-L8">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.KDPhenotypeRelationship" href="#Crispulator.KDPhenotypeRelationship"><code>Crispulator.KDPhenotypeRelationship</code></a> — <span class="docstring-category">Type</span></header><section><div><p>A type representing a relationship between degree of knockdown and effect on phenotype</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/library.jl#L22-L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.Cas9Behavior" href="#Crispulator.Cas9Behavior"><code>Crispulator.Cas9Behavior</code></a> — <span class="docstring-category">Type</span></header><section><div><p>A type representing the behavior of different Cas9s</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/library.jl#L52-L54">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.CRISPRn" href="#Crispulator.CRISPRn"><code>Crispulator.CRISPRn</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct CRISPRn &lt;: Crispulator.Cas9Behavior</code></pre><p>CRISPR KO behavior is more complex since sgRNA-directed DNA damage repair is stochastic. We assume that 2/3 of repair events at a given locus lead to a frameshift, and that the screen is carried out in diploid cells. The assumption that only bi-allelic frame-shift mutations lead to a phenotype in CRISPRn screens for most sgRNAs is supported by the empirical finding that in-frame deletions mostly do not show strong phenotypes, unless they occur in regions encoding conserved residues or domains<sup class="footnote-reference"><a id="citeref-2" href="#footnote-2">[2]</a></sup></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/library.jl#L64">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.CRISPRi" href="#Crispulator.CRISPRi"><code>Crispulator.CRISPRi</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct CRISPRi &lt;: Crispulator.Cas9Behavior</code></pre><p>CRISPRi behavior is simply determined by the activity of the guide</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/library.jl#L57">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.Delta" href="#Crispulator.Delta"><code>Crispulator.Delta</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct Delta &lt;: Distributions.Sampleable{Distributions.Univariate, Distributions.Discrete}</code></pre><p>Constructs a delta function at a given δ value. This distribution always emits the same value.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/library.jl#L277">source</a></section></article><h2 id="Miscellaneous-functions"><a class="docs-heading-anchor" href="#Miscellaneous-functions">Miscellaneous functions</a><a id="Miscellaneous-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Miscellaneous-functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Crispulator.signal" href="#Crispulator.signal"><code>Crispulator.signal</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">signal(guide_data; log2fc_col)
</code></pre><p>Computes the signal in an experiment, where the experimental signal is defined to be the average of signal of true hit genes. That value, the true hit gene signal, is the average ratio of the log2 fold change for a guide targeting a specific gene to the guide&#39;s theoretical phenotype.</p><p class="math-container">\[\frac{1}{N_{true} \times k} \sum_{i=1}^{N_{true}} \sum_{j=1}^k
\frac{\log_2 fc_{ij}}{\text{theo phenotype}_{ij}}\]</p><p>where <span>$N_{true}$</span> is the number of true hit genes and <span>$k$</span> is the number of genes.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/utils.jl#L143">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.noise" href="#Crispulator.noise"><code>Crispulator.noise</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">noise(guide_data; log2fc_col)
</code></pre><p>Computes the noise in an experiment, where noise is defined to be the standard deviation of log2 fold change in the negative controls. If more than 2 bins are used then the name of log2 fold change information can be provided to differentiate between multiple log 2 fold changes.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/utils.jl#L127">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.linear" href="#Crispulator.linear"><code>Crispulator.linear</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">linear(x, l)
</code></pre><p>Given value(s) <code>x</code> apply a simple linear function with maximum value <code>l</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/library.jl#L15">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.sigmoid" href="#Crispulator.sigmoid"><code>Crispulator.sigmoid</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">sigmoid(x, l, k, p)
</code></pre><p>Given value(s) <code>x</code> apply the sigmoidal function with maximum value <code>l</code>, a steepness <code>k</code>, and an inflection point <code>p</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/library.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.auprc" href="#Crispulator.auprc"><code>Crispulator.auprc</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">auprc(scores, classes, pos_labels; rev)
</code></pre><p>Computes the area under the Precision-Recall curve using a lower trapezoidal estimator, which is more accurate for skewed datasets.</p><p>K. Boyd, K. H. Eng, and C. D. Page, “Area under the Precision-Recall Curve: Point Estimates and Confidence Intervals,” in Machine Learning and Knowledge Discovery in Databases, H. Blockeel, K. Kersting, S. Nijssen, and F. Železný, Eds. Springer Berlin Heidelberg, 2013, pp. 451–466.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/utils.jl#L50">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.auroc" href="#Crispulator.auroc"><code>Crispulator.auroc</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">auroc(scores, classes, pos_labels; rev)
</code></pre><p>Optimized function for computing the area under the receiver operator characteristic curve.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/utils.jl#L13">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Crispulator.venn" href="#Crispulator.venn"><code>Crispulator.venn</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">venn(scores, classes, pos_labels; rev)
</code></pre><p>Given <code>N</code> positive examples, computes the percentage of the top <code>N/2</code> hits that are correct</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tlnagy/Crispulator.jl/blob/587d8cabc8e5b9879055147a44bab89821caa182/src/utils.jl#L103">source</a></section></article><section class="footnotes is-size-7"><ul><li class="footnote" id="footnote-1"><a class="tag is-link" href="#citeref-1">1</a>Kampmann M, Bassik MC, Weissman JS. Integrated platform for genome-wide screening and construction of high-density genetic interaction maps in mammalian cells. <em>Proc Natl Acad Sci U S A</em>. 2013;110:E2317–26.</li><li class="footnote" id="footnote-2"><a class="tag is-link" href="#citeref-2">2</a>Horlbeck MA, Gilbert LA, Villalta JE, Adamson B, Pak RA, Chen Y, Fields AP, Park CY, Corn JE, Kampmann M, Weissman JS: Compact and highly active next- generation libraries for CRISPR-mediated gene repression and activation. <em>Elife</em> 2016, 5.</li></ul></section></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../multiprocessing/">« Multiprocessing</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Tuesday 25 April 2023 19:20">Tuesday 25 April 2023</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
